test/test.media.ts

```diff
@@ -99,7 +99,7 @@ describe('Media', () => {
     const fileName = path.join(__dirname, '../../test/fixtures/mediabody.txt');
     const fileSize = (await pify(fs.stat)(fileName)).size;
     const google = new GoogleApis();
-    const youtube = google.youtube('v3');
+    const youtube = google.youtube<youtube_v3.Youtube>('v3');
```

Comment:
man, that type parameter makes things so much less readable :(

Reply:
Yeah :(  The issue here is a little complicated.  Because we offer more than one version of an API by default, I have to join the types if there are multiple.  If there's a single API, it's great.  If the multiple APIs have the same shape.. it's great!  When there are multiple versions and they're different... meh.  The only alternative I can think of here is to create a different API that references the specific typed API directly.

Reply:
Yeah, maybe it's possible to make `const youtube = google.youtube_v3` work. I see one more problem here, and it's huge: assume you have `v1` which works like `const something = google.something('v1')` but then API owners publish an incompatible `v2` that, if I read your explanation correctly, breaks the existing code as it will require `const something = google.something<something_v1.Something>('v1')` and the existing code will stop working. Is my understanding correct?

Reply:
Or did I understand your response wrong and it currently requires to use type parameter in all cases, even if there is just one version published?

Reply:
So - a few things.  Technically, the change to use these types was made a few commits ago.  The difference with *this* commit is that now instead of it just dictating the types of the response, it also dictates the types of the request.  

The problem you're pointing out here is very real.  There are a few options I can think of:

1. Make the generic type parameter required.  Today it defaults to `api_v1 | api_v2 | api_v3`. 

2. Make the generic type parameter default to `any`.  This would mean losing intellisense for anyone not specifically passing the type parameter.

3. Make the generic type parameter default to the latest semver version of the API.  This is *really* tricky.  But we could make our best guess of which generic type parameter to use.  The risk there is that the user chose a different version, and their type hinting is just flat out wrong. 

4. Introduce a new API for TypeScript users that doesn't rely on the version being passed in as a function argument. For example:

```js
const youtube = google.youtube.v3();
```

None of these are perfect, but I'm very interested in others thoughts :)

@google/google-node-team

Reply:
@JustinBeckwith Maybe you could use [overloads](https://www.typescriptlang.org/docs/handbook/functions.html#overloads) here? Here's a conceptual example (untested, I believe it is valid code):
```typescript
class youtube_v2 {
    _foo: number;
}
class youtube_v3 {
    _bar: number;
}
class Google {
    youtube(version: "v2"): youtube_v2;
    youtube(version: "v3"): youtube_v3;
    youtube(version: string): youtube_v3 | youtube_v2 {
        switch (version) {
            case "v2":
                return new youtube_v2();
            case "v3":
                return new youtube_v3();
            default:
                return new youtube_v3();
        }
    }
}

const g = new Google();
let v3 = g.youtube("v3"); // v3: youtube_v3
let v2 = g.youtube("v2"); // v2: youtube_v2
v3 = v2; // type error
```

Reply:
OK so I don't like (1) because it looks weird and the user needs to type more. I don't like (3) because it will make a possible semver minor update (adding new version is semver minor) incompatible as the default will change. I don't really care about (2) because I don't use intellisense :) but I understand why it's important. So, (4)? (as an alternative, `google.youtube_v3()` also LGTM)

Reply:
@acchou that's a really nifty idea, and for the case where you're just passing the version string I think it would work.  I wonder if this would handle the object literal case as well?

```js
const youtube = google.youtube({
  version: "v3",
  auth: myOAuth2Client
});
```

Reply:
I like what @acchou is suggesting here. Is that a possibility?

Reply:
+1 to what @acchou suggests. One thing I like about TypeScript is it can dispatch to a different function signature depending on a string constant value of a parameter; e.g. `EventEmitter`'s `on()` or `emit()` function, or like this case.

Reply:
@JustinBeckwith Indeed it does seem possible to handle object literals as well (reusing some definitions from above):
```typescript

interface GlobalOptions {
    url?: string;
    method?: string;
    baseURL?: string;
    auth?: any;
}
interface ServiceOptions extends GlobalOptions {
    version?: string;
}
interface ServiceOptionsv3 extends ServiceOptions {
    version: "v3";
}
interface ServiceOptionsv2 extends ServiceOptions {
    version: "v2";
}
class Google {
    youtube(version: "v2"): youtube_v2;
    youtube(version: "v3"): youtube_v3;
    youtube(options: ServiceOptionsv2): youtube_v2;
    youtube(options: ServiceOptionsv3): youtube_v3;
    youtube(options: string | ServiceOptions): youtube_v3 | youtube_v2 {
        let version = typeof options === "string" ? options : options.version;
        switch (version) {
            case "v2":
                return new youtube_v2();
            case "v3":
                return new youtube_v3();
            default:
                return new youtube_v3();
        }
    }
}

let o3 = g.youtube({ version: "v3" });
let o2 = g.youtube({ version: "v2" });
o3 = o2; // type error

let z3 = g.youtube({ version: "v3", auth: "foo" });
let z2 = g.youtube({ version: "v2", auth: "bar" });
z3 = z2; // type error
```

Reply:
Note that overloading with an exact object literal doesn't work:
```typescript
    youtube(options: { version: "v2" }): youtube_v2;
    youtube(options: { version: "v3" }): youtube_v3;
```
This will cause  type error in the z3 and z2 cases above, because the type is hard-coded to that specific object literal without allowing for any of the other keys in `GlobalOptions`. That's why we need `ServiceOptionsv2` and `ServiceOptionsv3`.

... and that means we may need to add two more overloads:
```typescript
    youtube(version: string): youtube_v3 | youtube_v2;
    youtube(options: ServiceOptions): youtube_v3 | youtube_v2;
```
It's debatable if these are needed. Without them, bugs like this are caught by the type checker:
```typescript
g.youtube("xyz");  // type error
g.youtube({ version: "xyz" }); // type error
```
The two extra overloads only seem to be needed if you expect to have a dynamically generated `version`. This could happen in a developer tool, for example. In that case, the code can cast the version or option to `any`.

My preference would be to leave the non-explicit version overloads out. It preserves a better experience and type checking for 99%+ of the users of the API.